File: mergesort.m
#import <Foundation/Foundation.h>

@interface MergeSort: NSObject

- (NSArray *) mergeSortArray:(NSArray *)unsortedArray;
@end

@implementation MergeSort

- (NSArray *) mergeArray:(NSArray *)leftArray rightArray:(NSArray *)rightArray {
    
    NSMutableArray *returnArray = [NSMutableArray array];
    int i = 0, e = 0;
    
    while (i < [leftArray count] && e < [rightArray count]) {
        int leftValue = [[leftArray objectAtIndex:i] intValue];
        int rightValue = [[rightArray objectAtIndex:e] intValue];
        if (leftValue < rightValue) {
            [returnArray addObject: [leftArray objectAtIndex:i++]];
        } else {
            [returnArray addObject: [rightArray objectAtIndex:e++]];
        }
    }
    
    while (i < [leftArray count]) {
        [returnArray addObject: [leftArray objectAtIndex:i++]];
    }
    
    while (e < [rightArray count]) {
        [returnArray addObject: [rightArray objectAtIndex:e++]];
    }
    
    return returnArray;
}


- (NSArray *) mergeSortArray:(NSArray *)unsortedArray {
    
    // Time complexity: Worst case is: O(n * log(n)).
    // Space complexity: O(3 * n) or O(n) due to the 3 NSArrays that are used.

    
    int count = (int)[unsortedArray count];
    if (count < 2) {
        return unsortedArray;
    }
    int middle = count / 2;
    NSArray *leftArray = [unsortedArray subarrayWithRange: NSMakeRange(0, middle)];
    NSArray *rightArray = [unsortedArray subarrayWithRange: NSMakeRange(middle, (count - middle))];
    
    NSArray *returnArray = [self mergeArray: [self mergeSortArray: leftArray] rightArray: [self mergeSortArray: rightArray]];
    return returnArray;
}

@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSArray *mergeSorttData = @[@4, @3, @10, @44, @6, @4, @1, @7, @15];
        
        MergeSort *ms = [[MergeSort alloc] init];
        // ms.cycles = 0;
        NSArray *mergeSortedArray = [ms mergeSortArray: mergeSorttData];
        NSLog(@"mergeSortedArray: %@", mergeSortedArray);
    }
    return 0;
}

File: mergesort.swift
func merge<T: Comparable>(_ leftArray: [T], _ rightArray: [T]) ->[T] {
    var sortedArray: [T] = []

    var rightIdx = 0
    var leftIdx = 0 

    while leftIdx < leftArray.count && rightIdx < rightArray.count {
        let left = leftArray[leftIdx]
        let right = rightArray[rightIdx]

        if left > right {
            sortedArray.append(right)
            rightIdx += 1
        }
        else if left < right {
            sortedArray.append(left)
            leftIdx += 1
        }
        else {
            sortedArray.append(left)
            leftIdx += 1 
            sortedArray.append(right)
            rightIdx += 1 
        }
    }
    while leftIdx < leftArray.count {
        sortedArray.append(leftArray[leftIdx])
        leftIdx += 1
    } 
    while rightIdx < rightArray.count {
        sortedArray.append(rightArray[rightIdx])
        rightIdx += 1
    }
    return sortedArray
}

func mergeSort<T: Comparable>(_ array: [T]) ->[T] {
    if array.count <= 1 {return array}
    let middleIdx = array.count/2

    let leftArray = mergeSort(Array(array[0..<middleIdx]))
    let rightArray = mergeSort(Array(array[middleIdx..<array.count]))
    return merge(leftArray, rightArray)
}

let test_1 = [5,9,1,3,4,6,6,3,2]
print(mergeSort(test_1))


